bound_query.go

```diff
@@ -0,0 +1,166 @@
+package bob
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/stephenafamo/bob/internal/mappings"
+)
+
+type MismatchedArgsError struct {
+	Expected int
+	Got      int
+}
+
+func (e MismatchedArgsError) Error() string {
+	return fmt.Sprintf("expected %d args, got %d", e.Expected, e.Got)
+}
+
+type DuplicateArgError struct{ Name string }
+
+func (e DuplicateArgError) Error() string {
+	return fmt.Sprintf("duplicate arg %s", e.Name)
+}
+
+type MissingArgError struct{ Name string }
+
+func (e MissingArgError) Error() string {
+	return fmt.Sprintf("missing arg %s", e.Name)
+}
+
+type binder[T any] interface {
+	ToArgs(T) []any
+}
+
+type BoundQuery[T any] struct {
+	query  []byte
+	binder binder[T]
+	start  int
+}
+
+func (b BoundQuery[T]) Bind(args T) BaseQuery[*cached] {
+	realArgs := b.binder.ToArgs(args)
+
+	return BaseQuery[*cached]{
+		Expression: &cached{
+			query: b.query,
+			args:  realArgs,
+			start: b.start,
+		},
+	}
+}
+
+func BindNamed[T any](q Query) (BoundQuery[T], error) {
+	return BindNamedN[T](q, 1)
```

Comment:
Maybe I could have access the the generated SQL query by returning it from here or having an accessor method in `BoundQuery`? This way I could potentially use it to prepare the query in the DB, and use `Build` just to get the parameter replacement.

Reply:
If you need to prepare the query in the DB and bind as necessary, you should use the `bob.Prepare`/`bob.PrepareQuery` methods. Those have also been modified to take the `Arg` struct.

Or am I misunderstanding the use case?